homeworks/solutions/torantie/3-smooth/src/halfedge.js

```diff
@@ -0,0 +1,632 @@
+
+        // FIXME: non-manifold edge count checking
+      }
+    }
+
+    // create boundary halfedges and hidden faces for the boundary
+    let hidx = indices.length
+    for (let i = 0; i < indices.length; i++) {
+      const he = this.halfedges[i]
+      if (hasTwin.get(he)) {
+        continue
+      }
+
+      // handle halfedge that has no twin
+      const f = new Face() // hidden face
+      let bcycle = []      // boundary cycle
+      let current = he
+      do {
+        const bhe = new Halfedge() // boundary halfedge
+        this.halfedges[hidx] = bhe
+        hidx++
+        bcycle.push(bhe)
+
+        // grab the next halfedge along the boundary that does not
+        // have a twin halfedge
+        let next = current.next
+        while (hasTwin.get(next)) {
+          next = next.twin.next
+        }
+
+        // set the current halfedge's attributes
+        bhe.vertex = next.vertex
+        bhe.edge = current.edge
+        bhe.onBoundary = true
+
+        // point the new halfedge and face to each other
+        bhe.face = f
+        f.halfedge = bhe
+
+        // point the new halfedge and twin to each other
+        bhe.twin = current
+        current.twin = bhe
+
+        current = next
+      } while(current != he)
+
+      // link the cycle of boundary halfedges together
+      const n = bcycle.length
+      for (let j = 0; j < n; j++) {
+        bcycle[j].next = bcycle[(j+n-1)%n]
+        bcycle[j].prev = bcycle[(j+1)%n]
+        hasTwin.set(bcycle[j], true)
+        hasTwin.set(bcycle[j].twin, true)
+      }
+    }
+
+    // reset indices
+    let index = 0
+    this.vertices.forEach(v => { v.idx = index++ })
+    index = 0
+    this.edges.forEach(e => { e.idx = index++ })
+    index = 0
+    this.faces.forEach(f => { f.idx = index++ })
+    index = 0
+    this.halfedges.forEach(h => { h.idx = index++ })
+
+    //Create Copy
+    for (const vertex of this.vertices) {
+      let vertexCopy = new Vertex()
+      vertexCopy.idx = vertex.idx
+      vertexCopy.halfedge = vertex.halfedge
+      vertexCopy.position = new Vector(vertex.position.x,vertex.position.y,vertex.position.z)
+      this.vertsOrig.push(vertexCopy)
+    }
+  }
+
+
+  laplaceMatrix(weightType) {
+    // TODO: construct the Laplace matrix.
+    const numberOfVertices = this.vertices.length
+    let weightTriplet = new Triplet(numberOfVertices, numberOfVertices)
+
+    for (const vert of this.vertices) {
+      const i = vert.idx
+      let sum = 1e-8 // Tikhonov regularization to get strict positive definite
+
+      vert.forEachHalfEdge(h => {
+        let w = 0
+
+        switch (weightType) {
+          case 'uniform':
+            w = 1
+            break
+          case 'cotan':
+            w = (h.cotan() + h.twin.cotan())/2
+        }
+
+        sum += w
+        //console.log("Value: "+w+" Position: ("+ i+","+h.twin.vertex.idx +")")
+        weightTriplet.addEntry(w, i, h.twin.vertex.idx)
+      })
+      weightTriplet.addEntry(-sum, i, i)
```

Comment:
Great that you stick to the comments below where we need compute `(I-tλL) f' = f`.

In the slides, I provided the Laplace matrix in a different way, thus we could compute `(I-tλL) ` as follows:

```
    const F = M.plus(W.timesReal(timeStep*smoothStep))
```